topcoffea/modules/HistEFT.py

```diff
@@ -192,7 +192,11 @@ def copy_sm(self):
     out._sumw = {}
     out._sumw2 = {}
     for sparse_key in self._sumw.keys():
-      out._sumw[sparse_key]=self._sumw[sparse_key][:,0]
+      is_eft_bin = (self._sumw[sparse_key].shape != self._dense_shape)
+      if is_eft_bin:
+        out._sumw[sparse_key]=self._sumw[sparse_key][:,0]
+      else:
+        out._sumw[sparse_key]=self._sumw[sparse_key]
```

Comment:
You might want to check whether or not L199 creates a fresh array or if it just saves a reference to an underlying array and whether that's intended or not. For example, if you take the return of this method and modified it in some way, does that propagate back to the original:
```python
bin_idx = 1 # or w/e might be a valid value
sparse_key = ("some","valid","key")

orig_hist._sumw[sparse_key][bin_idx] = 125
sm_copy = orig_hist.copy_sm()
sm_copy._sumw[sparse_key][bin_idx] = 42
assert(sm_copy._sumw[sparse_key][bin_idx] != orig_hist._sumw[sparse_key][bin_idx])
```

Reply:
@kmohrman sorry for taking so long to get around to testing this, but I think this isn't doing what you want it to do.

Below is a MRE that tries to showcase the problem:
```python
import numpy as np
from coffea import hist
from topcoffea.modules.HistEFT import HistEFT

nevts = 3
sparse_key = (hist.StringBin("ch01"),)
dense_idx  = 1 # Note: bin 0 is for the underflow
dense_vals = np.full(nevts,0.5) # [0.5 0.5 0.5]

orig_hist = HistEFT("events",[],hist.Cat("ch","ch"),hist.Bin("x","x-axis",1,0,1))
orig_hist.fill(ch=sparse_key[0],x=dense_vals)

# Check that we filled the histogram correctly
assert(orig_hist._sumw[sparse_key][dense_idx] == nevts) # --> OK

sm_copy = orig_hist.copy_sm()

# Check that the copied histogram has the correct bin value
assert(sm_copy._sumw[sparse_key][dense_idx] == nevts)   # --> OK

# Fill the copied histogram with more data
sm_copy.fill(ch=sparse_key[0],x=dense_vals)

assert(sm_copy._sumw[sparse_key][dense_idx] == 2*nevts) # --> OK

# Compare the bin content of the original and copied histogram
assert(sm_copy._sumw[sparse_key][dense_idx] != orig_hist._sumw[sparse_key][dense_idx])  # --> Fails
```

Basically, the underlying issue is that both `orig_hist` and `sm_copy` are storing a reference to the same underlying array. So when you modify the content of `sm_copy` it also modifies the content of `orig_hist`. I think this also impacts the eft case as well as showcased below:
```python
import numpy as np

orig_arr = np.array([[1,2,3],[4,5,6],[7,8,9]])
copy_arr = orig_arr[:,0]    # [1 4 7]
copy_arr[1] = 42

assert(copy_arr[1] != orig_arr[1,0])    # --> FAILS
```


I went and checked that changing:
```python
if is_eft_bin:
    out._sumw[sparse_key]=self._sumw[sparse_key][:,0]
else:
    out._sumw[sparse_key]=self._sumw[sparse_key]
```
to
```python
if is_eft_bin:
    out._sumw[sparse_key]=np.copy(self._sumw[sparse_key][:,0])
else:
    out._sumw[sparse_key]=np.copy(self._sumw[sparse_key])
```
seems to resolves the issue, but keep in mind that `np.copy` only does a "shallow" copy, so in the case where the "elements" of the `np.array` are themselves objects, then the copied array will still store a reference to those same objects. I _think_ this isn't an issue for us as all of the arrays in our histograms should just be simple numbers, but would be good to get others opinions on this if possible.

Reply:
fixed applied in commit `aefee7a`